iot_packet_verifier/src/pending.rs

```diff
@@ -0,0 +1,467 @@
+    pending_tables: &impl PendingTables,
+    solana: &S,
+) -> Result<(), ConfirmPendingError<S::Error>>
+where
+    S: SolanaNetwork,
+{
+    // Fetch all pending transactions and confirm them
+    let pending = pending_tables.fetch_all_pending_txns().await?;
+    for pending in pending {
+        // Sleep for at least a minute since the time of submission to
+        // give the transaction plenty of time to be confirmed
+        let time_since_submission = Utc::now() - pending.time_of_submission;
+        if Duration::minutes(1) > time_since_submission {
+            tokio::time::sleep((Duration::minutes(1) - time_since_submission).to_std()?).await;
+        }
+
+        let mut txn = pending_tables.begin().await?;
+        // We remove the transaction regardless of whether it has been confirmed
+        // or not:
+        txn.confirm_txn(&pending.signature).await?;
+        // Check if the transaction has been confirmed. If it has, remove the
+        // amount from the pending burns table
+        if solana
+            .confirm_transaction(&pending.signature)
+            .await
+            .map_err(ConfirmPendingError::SolanaError)?
+        {
+            txn.subtract_burned_amount(&pending.payer, pending.amount)
+                .await?;
+        }
+        // Commit our work.
+        txn.commit().await?;
+    }
+
+    Ok(())
+}
+
+#[async_trait]
+pub trait PendingTablesTransaction<'a> {
+    async fn confirm_txn(&mut self, signature: &Signature) -> Result<(), sqlx::Error>;
+
+    async fn subtract_burned_amount(
+        &mut self,
+        payer: &PublicKeyBinary,
+        amount: u64,
+    ) -> Result<(), sqlx::Error>;
+
+    async fn commit(self) -> Result<(), sqlx::Error>;
+}
+
+#[async_trait]
+impl PendingTables for PgPool {
+    type Transaction<'a> = Transaction<'a, Postgres>;
+
+    async fn begin<'a>(&'a self) -> Result<Self::Transaction<'a>, sqlx::Error> {
+        self.begin().await
+    }
+
+    async fn fetch_next_burn(&self) -> Result<Option<Burn>, sqlx::Error> {
+        sqlx::query_as(
+            "SELECT * FROM pending_burns WHERE amount >= $1 ORDER BY last_burn ASC LIMIT 1",
+        )
+        .bind(BURN_THRESHOLD)
+        .fetch_optional(self)
+        .await
+    }
+
+    async fn fetch_all_pending_burns(&self) -> Result<Vec<Burn>, sqlx::Error> {
+        sqlx::query_as("SELECT * FROM pending_burns")
+            .fetch_all(self)
+            .await
+    }
+
+    async fn fetch_all_pending_txns(&self) -> Result<Vec<PendingTxn>, sqlx::Error> {
+        sqlx::query_as("SELECT * FROM pending_txns")
+            .fetch_all(self)
+            .await
+    }
+
+    async fn submit_txn(
+        &self,
+        payer: &PublicKeyBinary,
+        amount: u64,
+        signature: &Signature,
+    ) -> Result<(), sqlx::Error> {
+        sqlx::query(
+            r#"
+            INSERT INTO pending_txns (signature, payer, amount, time_of_submission)
+            VALUES ($1, $2, $3, $4)
```

Comment:
is there any chance a txn could be submitted more than once that you need to have an `ON CONFLICT` here?

Reply:
I don't believe so, signatures should be unique even with the same inputs

Reply:
that's correct.. I'm assuming signature is enforced with a unique constraint?